### PR TITLE
feat: add assignees option to the copy and backport commands

### PR DESCRIPTION
**Pull request by v1v**
## Description of the change

Add assignees option to the copy and backport commands

## Related issues

Fix https://github.com/Mergifyio/mergify-engine/issues/2287

### Development

- [x] All checks must pass (Semantic Pull Request, pep8, requirements, unit tests, functional tests, security checks, …)
- [x] The code changed/added as part of this pull request must be covered with tests
- [ ] Hotfixes must have a link to Sentry issue and the ``hotfix`` label
- [ ] Features must have a link to a Linear task

### Code Review

Code review policies are handled and automated by Mergify.

* When all tests pass, reviewers will be assigned automatically.
* When change is approved by at least one review and no pending review are
  remaining, pull request is retested against its base branch.
* The pull request is then merged automatically.

### CODE REVIEW

**Comment by v1v** [comment]
Similar to https://github.com/Mergifyio/mergify-engine/pull/2334. 

I've no clue how to test this

**Review by sileht — COMMENTED** [review]
Maybe we can be more generic like `assign` action does, for example:

`voluptuous.Required("assignees", default[]: [types.Jinja2])`

And then the configuration will look like this:

```
backport:
    assignees: ["{{ author }}"]
```

What do you think?

File: docs/mergify_engine_gendoc.py [hunk 1]
```diff
@@ -63,11 +63,13 @@ def write_redirect(app, pagename, templatename, context, doctree):
...
     with open(os.path.join(app.outdir, pagename + ".html"), "w") as f:
-        f.write(f"""<head>
+        f.write(
+            f"""<head>
   <meta http-equiv="refresh" content="0; URL={url}">
   <link rel="canonical" href="{url}">
 </head>
-""")
+"""
+        )
```
Comment (v1v) [inline docs/mergify_engine_gendoc.py:72]: Unrelated change caused by the `tox -e black` command

**Comment by v1v** [comment]
I've just changed the implementation to use the suggestion from https://github.com/Mergifyio/mergify-engine/pull/2336#pullrequestreview-626849902

Though, I might need some help with the test data

**Comment by sileht** [comment]
Fixtures for the functional need to setup a GitHub app and a couple of GitHub accounts and API keys.

I will take care of it.

File: mergify_engine/actions/__init__.py [hunk 2]
```diff
@@ -120,3 +120,19 @@ async def get_rule(
         return EvaluatedActionRule(
             "", rules.RuleConditions([]), rules.RuleMissingConditions([])
         )
+
+    async def wanted_users(
```
Comment (sileht) [inline mergify_engine/actions/__init__.py:124]: Method should be static.

File: mergify_engine/duplicate_pull.py [hunk 3]
```diff
@@ -349,4 +351,10 @@ async def duplicate(
             json={"labels": effective_labels},
         )
 
+    if assignees is not None:
```
Comment (sileht) [inline mergify_engine/duplicate_pull.py:354]: Need also checks `len(assignes) > 0` to avoid useless HTTP request

File: docs/mergify_engine_gendoc.py (hunk 1 above)
Comment (sileht) [reply docs/mergify_engine_gendoc.py:72]: I will fix the tox target

**Review by sileht — COMMENTED** [review]
I just post some remarks, but I have already fixed them locally. You don't have to fix them.

I'm currently generating the fixtures.

**Review by sileht — APPROVED** [review]

File: docs/source/actions/backport.rst [hunk 4]
```diff
@@ -70,7 +70,13 @@ Options
      - ``conflicts``
      - The label to add to the created pull request if it has conflicts and
        ``ignore_conflicts`` is set to ``true``.
+   * - ``assignees``
+     - list of :ref:`data type template`
+     -
+     - The users to assign to the pull request.
 
+The list of users in ``assignees`` is based on :ref:`data type template`, you can use
+e.g. ``{{author}}`` to assign the pull request to its author.
```
Comment (jd) [inline docs/source/actions/backport.rst:79]:
```suggestion
     - Users to assign the newly created pull request. As the type is :ref:`data type template`, you could use, e.g., ``{{author}}`` to assign the pull request to its original author.
```

File: docs/source/actions/backport.rst [hunk 5]
```diff
@@ -94,5 +100,6 @@ added and the pull request merged:
           backport:
             branches:
               - stable
+            assignees: ["{{ author }}"]
```
Comment (jd) [inline docs/source/actions/backport.rst:103]:
```suggestion
            assignees:
              - "{{ author }}"
```
I feel it's more readable :)

File: docs/source/actions/copy.rst [hunk 6]
```diff
@@ -51,6 +51,13 @@ Options
      - ``conflicts``
      - The label to add to the created pull request if it has conflicts and
        ``ignore_conflicts`` is set to ``true``.
+   * - ``assignees``
+     - list of :ref:`data type template`
+     -
+     - The users to assign to the pull request.
+
+The list of users in ``assignees`` is based on :ref:`data type template`, you can use
+e.g. ``{{author}}`` to assign the pull request to its author.
```
Comment (jd) [inline docs/source/actions/copy.rst:60]: Change like backport above.

**Review by jd — DISMISSED** [review]
Just small doc change

**Review by jd — APPROVED** [review]
